### PR TITLE
External auth and settings

### PR DESCRIPTION
**Pull request by darh**
This PR changes:
 - moves `/oidc`endpoints  into `/auth/external/openid-connect-....`
 - moves `/social/{provider}` endpoints into /auth/external/{provider}`
 - introduces internal/settings package for settings management (it is db-table agnostic, does not care about permissions, provides a set of helper functions & types to ease settings management)
 - introduces `system/internal/settings` - wrapper for `internal/settings`
 - introduces `system` `/settings` endpoints (messagin, crm/compose will follow the same priciple)
 - all cli code is moved out of `cmd/system-cli` to `system/cli`
 - added settings management cli commands
 - all four (standard) providers are now configured from (db) settings
 - support for arbitray number of OIDC providers

Todo:
 - [x] add permission check under system/internal/settings
 - [x] write instructions on how to migrate from .env to settings for external auth
 - [x] write instructions how to configure external auth on first deploy
 - [x] additonal checks & testing for all external auth flows
 - [x] make sure user email is always unique
 - [x] remove obsolete ("social") config/flags
 - [x] fix empty credential types

Future:
  - [ ] ability to enable/disable new user registration (only pre-created users)
 - [ ] ability to enable/disable "local" user
  - [ ] support for local users
 - [ ] add ability to read & apply new external-auth settings without restarting the whole service

### CODE REVIEW

File: system/internal/service/user.go [hunk 1]
```diff
@@ -121,6 +158,7 @@ func (svc *user) FindOrCreate(user *types.User) (out *types.User, err error) {
 		out, err = svc.user.FindBySatosaID(user.SatosaID)
 
 		if err == repository.ErrUserNotFound {
+			// @todo do we allow autocreation of nonexisting users?
```
Comment (titpetric) [inline system/internal/service/user.go]: This is weird, no.

File: system/internal/service/user.go [hunk 2]
```diff
@@ -142,10 +179,14 @@ func (svc *user) FindOrCreate(user *types.User) (out *types.User, err error) {
 
 func (svc *user) Create(input *types.User) (out *types.User, err error) {
 	return out, svc.db.Transaction(func() error {
-		// Encrypt user password
 		if out, err = svc.user.Create(input); err != nil {
 			return err
 		}
+
+		if !svc.prm.CanCreateUser() {
+			return errors.New("not allowed to create users")
```
Comment (titpetric) [inline system/internal/service/user.go]: This should be moved up before actually creating the user record.

File: internal/config/social.go [hunk 3]
```diff
@@ -11,6 +11,9 @@ type (
 	Social struct {
 		Enabled bool
 
+		OidcKey        string
```
Comment (titpetric) [inline internal/config/social.go]: Oidc should be OIDC, much like HTTP

Comment (darh) [reply internal/config/social.go]: this (config/social) is obsolete, will add to todo

File: system/internal/service/user.go (hunk 2 above)
Comment (darh) [reply system/internal/service/user.go]: True. Mea culpa.

File: system/internal/service/user.go (hunk 1 above)
Comment (darh) [reply system/internal/service/user.go]: This (logic) is still needed but ONLY in context of external logins, where we create users on-the-fly on successful (external) authentication. Auth service already does that.

Removed from user service, added comment to auth service.

File: system/internal/auth/social/main.go [hunk 4]
```diff
@@ -2,8 +2,9 @@ package social
 
 import (
 	"github.com/crusttech/crust/internal/config"
+	"github.com/crusttech/crust/internal/settings"
 )
 
-func Init(c *config.Social) {
-	initGoth(c)
+func Init(c *config.Social, finder settings.Finder) {
```
Comment (titpetric) [inline system/internal/auth/social/main.go]: You shouldn't use `main.go` as the filename outside of cmd/ (use `init.go` or `setup.go` or something more apt)

Comment (darh) [reply system/internal/auth/social/main.go]: ack.

**Pull request closed without merging**